File: tinyredis-remote/src/main/java/com/github/cjqcn/tinyredis/remote/RedisServerHandler.java
/*
 * Copyright 2012 The Netty Project
 *
 * The Netty Project licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
package com.github.cjqcn.tinyredis.remote;

import io.netty.channel.ChannelHandler.Sharable;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.handler.codec.redis.SimpleStringRedisMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Handler implementation for the echo server.
 */
@Sharable
public class RedisServerHandler extends ChannelInboundHandlerAdapter {

    private static final Logger LOG = LoggerFactory.getLogger(RedisServerHandler.class);

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) {

        LOG.debug("{}", msg);

        SimpleStringRedisMessage simpleStringRedisMessage = new SimpleStringRedisMessage("pong");
        ctx.write(simpleStringRedisMessage);
    }

    @Override
    public void channelReadComplete(ChannelHandlerContext ctx) {
        ctx.flush();
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        // Close the connection when an exception is raised.
        cause.printStackTrace();
        ctx.close();
    }

}

File: tinyredis-remote/src/main/java/com/github/cjqcn/tinyredis/remote/RedisServerHandler.kt
package com.github.cjqcn.tinyredis.remote

import com.github.cjqcn.tinyredis.core.client.RedisClient
import com.github.cjqcn.tinyredis.core.command.RedisCommand
import com.github.cjqcn.tinyredis.core.command.impl.GetCommand
import com.github.cjqcn.tinyredis.core.command.impl.SetCommand
import com.github.cjqcn.tinyredis.core.server.RedisServer
import com.github.cjqcn.tinyredis.core.struct.StringObject
import io.netty.channel.ChannelHandler
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.handler.codec.CodecException
import io.netty.handler.codec.redis.*
import io.netty.util.CharsetUtil
import org.slf4j.LoggerFactory

@ChannelHandler.Sharable
class RedisServerHandler(redisServer: RedisServer) : ChannelInboundHandlerAdapter() {
    companion object {
        private val LOGGER = LoggerFactory.getLogger(RedisServerHandler::class.java)
    }

    private val redisClient: RedisClient = redisServer.createClient();

    override fun handlerRemoved(ctx: ChannelHandlerContext?) {
        super.handlerRemoved(ctx)
        redisClient.destroy()
    }

    override fun channelRead(ctx: ChannelHandlerContext, msg: Any?) {
        if (msg !is RedisMessage) {
            return
        }
        val command = decodeRedisCommand(msg)
        val response = command?.execute()
        ctx.write(SimpleStringRedisMessage(response?.decode() ?: "NON"))
    }

    override fun channelReadComplete(ctx: ChannelHandlerContext) {
        ctx.flush()
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        cause.printStackTrace()
        ctx.close()
    }

    private fun decodeRedisCommand(msg: RedisMessage): RedisCommand? {
        if (msg !is ArrayRedisMessage) {
            return null
        }
        val first = msg.children()[0];
        if (first !is FullBulkStringRedisMessage) {
            return null
        }
        when (getString(first)) {
            "get" -> return GetCommand(redisClient, StringObject(decodeToString(msg.children()[1])))
            "set" -> return SetCommand(redisClient, StringObject(decodeToString(msg.children()[1])),
                    StringObject(decodeToString(msg.children()[2])), 10000L)
        }
        return null
    }

    private fun decodeToString(msg: RedisMessage): String {
        return when (msg) {
            is SimpleStringRedisMessage -> {
                msg.content()
            }
            is ErrorRedisMessage -> {
                msg.content()
            }
            is IntegerRedisMessage -> {
                msg.value().toString()
            }
            is FullBulkStringRedisMessage -> {
                getString(msg)
            }
            else -> {
                throw CodecException("unknown message type: $msg")
            }
        }
    }

    private fun getString(msg: FullBulkStringRedisMessage): String {
        return if (msg.isNull) {
            "(null)"
        } else msg.content().toString(CharsetUtil.UTF_8)
    }
}